File: backend/app/schemas/analytics.py
"""Esquemas del dominio Analítico (Motor de Toma de Decisiones).

Espejo tipado de lo que devuelven las funciones SQL de agregación. La capa de
presentación (panel de admin) consume `AnalyticsDashboard` en una sola llamada.
"""

from datetime import date

from pydantic import BaseModel, Field


class PuntoAsignacion(BaseModel):
    """Minutos promedio de espera en 'buscando_conductor' por hora del día."""

    hora: int = Field(..., ge=0, le=23, description="Hora del día (0-23)")
    minutos_promedio: float = Field(..., description="Espera promedio hasta 'asignado'")
    muestras: int = Field(..., description="Cantidad de fletes que aportan al promedio")


class PuntoSLAKyc(BaseModel):
    """Horas promedio que tarda la validación KYC, por semana de ingreso."""

    semana: date = Field(..., description="Lunes de la semana (date_trunc 'week')")
    horas_promedio: float = Field(..., description="Tiempo de verificación en horas")
    muestras: int = Field(..., description="Conductores verificados esa semana")


class ResumenOperativo(BaseModel):
    """KPIs inteligentes de cabecera."""

    fletes_completados: int = 0
    fletes_cancelados: int = 0
    tasa_conversion: float = Field(0, description="% entregados sobre entregados+cancelados")
    comision_proyectada: float = Field(0, description="10% del monto de fletes entregados (CLP)")


class AnalyticsDashboard(BaseModel):
    """Payload completo que alimenta el Dashboard BI en una sola petición."""

    resumen: ResumenOperativo
    sla_kyc_semanal: list[PuntoSLAKyc]
    asignacion_por_hora: list[PuntoAsignacion]

File: frontend/src/lib/analytics.ts
/**
 * Capa de API del Motor Analítico (Dashboard BI).
 *
 * Lee las métricas agregadas desde el backend protegido por rol 'administrador',
 * adjuntando el JWT de la sesión de Supabase. La agregación abarca TODA la
 * operación (sin RLS), por eso vive detrás del endpoint de admin y NO se
 * consulta a Supabase directamente desde el panel.
 */

import { supabase } from './supabase'

const API_URL =
  import.meta.env.VITE_API_URL ?? import.meta.env.VITE_BACKEND_URL ?? 'http://localhost:8000'

/** Espejo de PuntoAsignacion (analytics.py). */
export interface PuntoAsignacion {
  hora: number
  minutos_promedio: number
  muestras: number
}

/** Espejo de PuntoSLAKyc (analytics.py). */
export interface PuntoSLAKyc {
  semana: string // ISO date (YYYY-MM-DD)
  horas_promedio: number
  muestras: number
}

/** Espejo de ResumenOperativo (analytics.py). */
export interface ResumenOperativo {
  fletes_completados: number
  fletes_cancelados: number
  tasa_conversion: number
  comision_proyectada: number
}

/** Espejo de AnalyticsDashboard (analytics.py). */
export interface AnalyticsDashboard {
  resumen: ResumenOperativo
  sla_kyc_semanal: PuntoSLAKyc[]
  asignacion_por_hora: PuntoAsignacion[]
}

async function authHeaders(): Promise<Record<string, string>> {
  const { data } = await supabase.auth.getSession()
  const token = data.session?.access_token
  return token ? { Authorization: `Bearer ${token}` } : {}
}

/** Trae el payload completo del Dashboard BI en una sola petición. */
export async function getDashboardBI(): Promise<AnalyticsDashboard> {
  const res = await fetch(`${API_URL}/api/v1/analytics/dashboard`, {
    headers: { Accept: 'application/json', ...(await authHeaders()) },
  })
  if (!res.ok) throw new Error('No se pudieron cargar las métricas analíticas.')
  return (await res.json()) as AnalyticsDashboard
}
